ydb/query/transaction.py

```diff
@@ -0,0 +1,338 @@
+
+    def __exit__(self, *args, **kwargs):
+        """
+        Closes a transaction context manager and rollbacks transaction if
+        it is not rolled back explicitly
+        """
+        if self._tx_state.tx_id is not None:
+            # It's strictly recommended to close transactions directly
+            # by using commit_tx=True flag while executing statement or by
+            # .commit() or .rollback() methods, but here we trying to do best
+            # effort to avoid useless open transactions
+            logger.warning("Potentially leaked tx: %s", self._tx_state.tx_id)
+            try:
+                self.rollback()
+            except issues.Error:
+                logger.warning("Failed to rollback leaked tx: %s", self._tx_state.tx_id)
+
+            self._tx_state.tx_id = None
+
+    @property
+    def session_id(self):
+        """
+        A transaction's session id
+
+        :return: A transaction's session id
+        """
+        return self._session_state.session_id
+
+    @property
+    def tx_id(self):
+        """
+        Returns a id of open transaction or None otherwise
+
+        :return: A id of open transaction or None otherwise
+        """
+        return self._tx_state.tx_id
+
+    def begin(self, settings=None):
+        """
+        Explicitly begins a transaction
+
+        :param settings: A request settings
+
+        :return: An open transaction
+        """
+        self._tx_state._check_invalid_transition(QueryTxStateEnum.BEGINED)
+
+        return self._driver(
+            _create_begin_transaction_request(self._session_state, self._tx_state),
+            _apis.QueryService.Stub,
+            _apis.QueryService.BeginTransaction,
+            wrap_tx_begin_response,
+            settings,
+            (self._session_state, self._tx_state, self),
+        )
+
+    def commit(self, settings=None):
+        """
+        Calls commit on a transaction if it is open otherwise is no-op. If transaction execution
+        failed then this method raises PreconditionFailed.
+
+        :param settings: A request settings
+
+        :return: A committed transaction or exception if commit is failed
+        """
+        if self._tx_state._already_in(QueryTxStateEnum.COMMITTED):
+            return
+        self._tx_state._check_invalid_transition(QueryTxStateEnum.COMMITTED)
+
+        self._ensure_prev_stream_finished()
+
+        return self._driver(
+            _create_commit_transaction_request(self._session_state, self._tx_state),
+            _apis.QueryService.Stub,
+            _apis.QueryService.CommitTransaction,
+            wrap_tx_commit_response,
+            settings,
+            (self._session_state, self._tx_state, self),
+        )
+
+    def rollback(self, settings=None):
+        if self._tx_state._already_in(QueryTxStateEnum.ROLLBACKED):
+            return
+
+        self._tx_state._check_invalid_transition(QueryTxStateEnum.ROLLBACKED)
+
+        self._ensure_prev_stream_finished()
```

Comment:
What about not read full stream on rollback?
we need only: no commit + optional sucessfully rollback, no need result or execute any queries